presto-spi/src/main/java/com/facebook/presto/spi/function/ComplexTypeFunctionDescriptorFunctionDescriptor.java

```diff
@@ -0,0 +1,172 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spi.function;
+
+import com.facebook.presto.common.Subfield;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static com.facebook.presto.common.Subfield.allSubscripts;
+import static java.util.Collections.emptyList;
+import static java.util.Collections.emptySet;
+import static java.util.Collections.unmodifiableList;
+import static java.util.Collections.unmodifiableSet;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Contains properties that describe how the function operates on Map or Array inputs.
+ */
+public class ComplexTypeFunctionDescriptorFunctionDescriptor
+{
+    public static final ComplexTypeFunctionDescriptorFunctionDescriptor DEFAULT = new ComplexTypeFunctionDescriptorFunctionDescriptor(
+            true,
+            true,
+            true,
+            emptyList(),
+            Optional.of(emptySet()),
+            Optional.empty());
+
+    /**
+     * Indicates whether the function internally accessing subfields. This does not include the subfield access through the functional parameter (lambda) passed as an
```

Comment:
>what it means that this does not include subfield access via lambda?

Function `filter` does not access the subfields internally even though while evaluating the function it accesses the subfields inside lambda parameter. 

>why not?

Because we will go inside the lambda and collect all subfields that were accessed there.